### PR TITLE
Polish up APISIX metrics to follow Service concept 

### PR DESCRIPTION
**Pull request by pg-yang**
- [ ] If this pull request closes/resolves/fixes an existing issue, replace the issue number. Closes #<issue number>.
- [ ] Update the [`CHANGES` log](https://github.com/apache/skywalking/blob/master/docs/en/changes/changes.md).

All discussion could be found here  https://github.com/apache/skywalking-showcase/pull/83#issuecomment-1272518463

### CODE REVIEW

**Comment by pg-yang** [comment]
image

image

image

File: docs/en/setup/backend/backend-apisix-monitoring.md [hunk 1]
```diff
@@ -13,33 +13,68 @@ SkyWalking leverages OpenTelemetry Collector to transfer the metrics to
 2. Set up [OpenTelemetry Collector ](https://opentelemetry.io/docs/collector/getting-started/#docker). For details on Prometheus Receiver in OpenTelemetry Collector, refer to [here](../../../../test/e2e-v2/cases/apisix/otel-collector/otel-collector-config.yaml).
 3. Config SkyWalking [OpenTelemetry receiver](opentelemetry-receiver.md).
 
-### MySQL Monitoring
+### APISIX Monitoring
 [APISIX prometheus plugin](https://apisix.apache.org/docs/apisix/plugins/prometheus/) provide multiple dimensions metrics for APISIX server , node , route , etc. 
-Accordingly, Skywalking observes the status, payload, and latency of the APISIX server, which is cataloged as a `LAYER: APISIX` `Service` in the OAP. Meanwhile, the server node would be recognized as `instance, and the route rule would be recognized as `endpoint`.
+Accordingly, Skywalking observes the status, payload, and latency of the APISIX server, which is cataloged as a `LAYER: APISIX` `Service` in the OAP. Meanwhile, the APISIX instance would be recognized as `LAYER: APISIX` `instance`, and the route rule , node would be recognized as `LAYER: APISIX` `endpoint`.
```
Comment (wu-sheng) [inline docs/en/setup/backend/backend-apisix-monitoring.md:18]:
```suggestion
Accordingly, Skywalking observes the status, payload, and latency of the APISIX server, which is cataloged as a `LAYER: APISIX` `Service` in the OAP. Meanwhile, the instances would be recognized as `LAYER: APISIX` `instance`s. The route rules and nodes would be recognized as `endpoint`s with `route/` and `node/` prefixes.
```

File: docs/en/setup/backend/backend-apisix-monitoring.md [hunk 2]
```diff
@@ -13,33 +13,68 @@ SkyWalking leverages OpenTelemetry Collector to transfer the metrics to
...
+
+#### Specify Skywalking Service name  
+
+Skywalking extract OTEL Collector attribute `skywalking_service` as `Service` name .
```
Comment (wu-sheng) [inline docs/en/setup/backend/backend-apisix-monitoring.md:22]:
```suggestion
SkyWalking expects OTEL Collector attribute `skywalking_service` to be the `Service` name.
```

File: docs/en/setup/backend/backend-apisix-monitoring.md [hunk 3]
```diff
@@ -13,33 +13,68 @@ SkyWalking leverages OpenTelemetry Collector to transfer the metrics to
...
+
+You could specify `skywalking_service` attribute by `labels` keyword if using Prometheus receiver , as following :
```
Comment (wu-sheng) [inline docs/en/setup/backend/backend-apisix-monitoring.md:24]:
```suggestion
Make sure `skywalking_service` attribute exists through `static_configs` of OTEL Prometheus scape config.
```

File: docs/en/setup/backend/backend-apisix-monitoring.md [hunk 4]
```diff
@@ -13,33 +13,68 @@ SkyWalking leverages OpenTelemetry Collector to transfer the metrics to
...
+
+```yaml      
+processors:
+  resource/skywalking-service:
+    attributes:
+    - key: skywalking_service   
+      value: exmple_service_name # Specify Skywalking Service name 
+      action: insert                     
+```
+Note , if you don't specify `service` attribute , Skywalking would use `APISIX` as default service name
```
Comment (wu-sheng) [inline docs/en/setup/backend/backend-apisix-monitoring.md:48]:
```suggestion
Notice , if you don't specify `service` attribute, SkyWalking OAP would use `APISIX` as the default service name  
```

File: oap-server/server-starter/src/main/resources/otel-rules/apisix.yaml [hunk 5]
```diff
@@ -29,53 +29,72 @@
...
+  - name: endpoint_http_latency
+    exp: apisix_http_latency.tagNotEqual('route','').tag({tags->tags.route = 'route::'+tags['route']}).sum(['type','le','service_name','route']).histogram().histogram_percentile([50,70,90,99]).endpoint(['service_name'],['route'], Layer.APISIX)
+    # Reorganization metrics which has `node` label as endpoint , that is formatted to `node/{node}`
+  - name: endpoint_http_status
+    exp: apisix_http_status.tagNotEqual('node','').tag({tags->tags.node = 'node::'+tags['node']}).sum(['code','service_name','node']).rate('PT1M').endpoint(['service_name'],['node'], Layer.APISIX)
+  - name: endpoint_bandwidth
+    exp: apisix_bandwidth.tagNotEqual('node','').tag({tags->tags.node = 'node::'+tags['node']}).sum(['type','service_name','node']).rate('PT1M').endpoint(['service_name'],['node'], Layer.APISIX)
+  - name: endpoint_http_latency
+    exp: apisix_http_latency.tagNotEqual('node','').tag({tags->tags.node = 'node::'+tags['node']})sum(['type','le','service_name','node']).histogram().histogram_percentile([50,70,90,99]).endpoint(['service_name'],['node'], Layer.APISIX)
```
[the remark this answers is not in the document]
Comment (wu-sheng) [reply oap-server/server-starter/src/main/resources/otel-rules/apisix.yaml:100]: What do you mean?

File: docs/en/setup/backend/backend-apisix-monitoring.md (hunk 4 above)
Comment (pg-yang) [reply docs/en/setup/backend/backend-apisix-monitoring.md:48]: `skywalking_service` attribute

Comment (wu-sheng) [reply docs/en/setup/backend/backend-apisix-monitoring.md:48]: This is the original version issue.

File: oap-server/server-starter/src/main/resources/otel-rules/apisix.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (pg-yang) [reply oap-server/server-starter/src/main/resources/otel-rules/apisix.yaml:100]:
```
  - name: endpoint_http_status
    exp: apisix_http_status.tagNotEqual('route','').tag({tags->tags.route = 'route/'+tags['route']}).sum(['code','service_name','route']).rate('PT1M').endpoint(['service_name'],['route'], Layer.APISIX)
  - name: endpoint_http_status
    exp: apisix_http_status.tagNotEqual('node','').tag({tags->tags.node = 'node/'+tags['node']}).sum(['code','service_name','node']).rate('PT1M').endpoint(['service_name'],['node'], Layer.APISIX)
```

In order to make both of  them use one metrics(table)  , I  use  same name .

**Comment by pg-yang** [comment]
> `route::` and `node::` should be updated as `route/` and `node/`. Because in ElasticSearch and BanyanDB search, it relies on word splitter, `/` is in that list, `::` is not.

Update  : 

image

**Comment by wu-sheng** [comment]
One more suggestion, you could set metrics for endpoint list for now.

**Comment by pg-yang** [comment]
> 

Like this ?
image

**Comment by wu-sheng** [comment]
Yes.

**Comment by wu-sheng** [comment]
Is `upstream` better than `node`? Is the IP upstream addr?

**Comment by pg-yang** [comment]
> Is `upstream` better than `node`? Is the IP upstream addr?

I have update the tile , keep default value . IP is apisix instance . 

image

**Comment by wu-sheng** [comment]
What do you mean? Isn't 172 a remote and upstream addr?

**Comment by pg-yang** [comment]
> 

Sorry  , Let's  change to upstream

**Comment by pg-yang** [comment]
> upstream

My  metrics is  labeled  , and  other similar ui-template is  not labeled .  Couldn't  support labeled metrics ?

**Comment by wu-sheng** [comment]
@pg-yang Let's continue without this.

@Fine0830 Could we support labeled value? I suppose we need an issue to discuss how to show brief values if supported.

File: oap-server/server-starter/src/main/resources/otel-rules/apisix.yaml [hunk 6]
```diff
@@ -29,53 +29,72 @@
 #    "-P-6H+3M"  -- parses as "+6 hours and -3 minutes"
 # </pre>
 filter: "{ tags -> tags.job_name == 'apisix-monitoring' }" # The OpenTelemetry job name
-expPrefix:  tag({tags -> tags.host_name = 'apisix::' + tags.service_instance_id})
+expPrefix: tag({tags -> tags.service_name = 'APISIX::'+(tags['skywalking_service']?.trim()?tags['skywalking_service']:'APISIX')})
```
Comment (kezhenxu94) [inline oap-server/server-starter/src/main/resources/otel-rules/apisix.yaml:32]: Try this

```suggestion
expPrefix: tag({tags -> tags.service_name = 'APISIX::'+(tags['skywalking_service']?.trim() ?: 'APISIX')})
```

**Comment by wu-sheng** [comment]
Do we need test updates due to endpoint format changed?

**Comment by pg-yang** [comment]
It's time to dinner . Hot pot is  great 

> Do we need test updates due to endpoint format changed?

Needn't  , E2E only check whether endpoint has value  rather than specific value 




image

**Review by wu-sheng — DISMISSED** [review]

File: docs/en/setup/backend/backend-apisix-monitoring.md [hunk 7]
```diff
@@ -13,33 +13,68 @@ SkyWalking leverages OpenTelemetry Collector to transfer the metrics to
...
 3. Config SkyWalking [OpenTelemetry receiver](opentelemetry-receiver.md).
 
-### MySQL Monitoring
-[APISIX prometheus plugin](https://apisix.apache.org/docs/apisix/plugins/prometheus/) provide multiple dimensions metrics for APISIX server , node , route , etc. 
-Accordingly, Skywalking observes the status, payload, and latency of the APISIX server, which is cataloged as a `LAYER: APISIX` `Service` in the OAP. Meanwhile, the server node would be recognized as `instance, and the route rule would be recognized as `endpoint`.
+### APISIX Monitoring
+[APISIX prometheus plugin](https://apisix.apache.org/docs/apisix/plugins/prometheus/) provide multiple dimensions metrics for APISIX server , upstream , route , etc. 
+Accordingly, Skywalking observes the status, payload, and latency of the APISIX server, which is cataloged as a `LAYER: APISIX` `Service` in the OAP. Meanwhile, the instances would be recognized as `LAYER: APISIX` `instance`s. The route rules and nodes would be recognized as `endpoint`s with `route/` and `upstream/` prefixes.
```
Comment (kezhenxu94) [inline docs/en/setup/backend/backend-apisix-monitoring.md:18]:
```suggestion
Accordingly, SkyWalking observes the status, payload, and latency of the APISIX server, which is cataloged as a `LAYER: APISIX` `Service` in the OAP. Meanwhile, the instances would be recognized as `LAYER: APISIX` `instance`s. The route rules and nodes would be recognized as `endpoint`s with `route/` and `upstream/` prefixes.
```

File: docs/en/setup/backend/backend-apisix-monitoring.md [hunk 8]
```diff
@@ -13,33 +13,68 @@ SkyWalking leverages OpenTelemetry Collector to transfer the metrics to
...
+
+#### Specify Skywalking Service name
```
Comment (kezhenxu94) [inline docs/en/setup/backend/backend-apisix-monitoring.md:20]:
```suggestion
#### Specify SkyWalking Service name  
```

**Review by kezhenxu94 — APPROVED** [review]
LGTM overall

**Comment by pg-yang** [comment]
ping 🔢

**Review by wu-sheng — APPROVED** [review]

**Pull request merged**